File: 2025-09-September-GFG-POTD/September 28 - Longest Bounded-Difference Subarray/Solution.py
from collections import deque

class Solution:
    def longestSubarray(self, arr, x):
        n = len(arr)
        minDeque, maxDeque = deque(), deque()
        
        left = 0
        bestLen, bestStart = 0, 0
        
        for right in range(n):
            # Maintain decreasing deque for max
            while maxDeque and arr[maxDeque[-1]] < arr[right]:
                maxDeque.pop()
            maxDeque.append(right)
            
            # Maintain increasing deque for min
            while minDeque and arr[minDeque[-1]] > arr[right]:
                minDeque.pop()
            minDeque.append(right)
            
            # Shrink window if difference exceeds x
            while arr[maxDeque[0]] - arr[minDeque[0]] > x:
                if maxDeque[0] == left:
                    maxDeque.popleft()
                if minDeque[0] == left:
                    minDeque.popleft()
                left += 1
            
            # Update best window
            if right - left + 1 > bestLen:
                bestLen = right - left + 1
                bestStart = left
        
        return arr[bestStart:bestStart + bestLen]

File: 2025-09-September-GFG-POTD/September 28 - Longest Bounded-Difference Subarray/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> longestSubarray(vector<int>& arr, int x) {
        int n = arr.size();
        deque<int> minDeque, maxDeque;
        
        int left = 0, bestLen = 0, bestStart = 0;
        
        for (int right = 0; right < n; right++) {
            // Maintain decreasing deque for max
            while (!maxDeque.empty() && arr[maxDeque.back()] < arr[right]) {
                maxDeque.pop_back();
            }
            maxDeque.push_back(right);
            
            // Maintain increasing deque for min
            while (!minDeque.empty() && arr[minDeque.back()] > arr[right]) {
                minDeque.pop_back();
            }
            minDeque.push_back(right);
            
            // Shrink window if difference exceeds x
            while (arr[maxDeque.front()] - arr[minDeque.front()] > x) {
                if (maxDeque.front() == left) maxDeque.pop_front();
                if (minDeque.front() == left) minDeque.pop_front();
                left++;
            }
            
            // Update best window
            if (right - left + 1 > bestLen) {
                bestLen = right - left + 1;
                bestStart = left;
            }
        }
        
        vector<int> result;
        for (int i = bestStart; i < bestStart + bestLen; i++) {
            result.push_back(arr[i]);
        }
        return result;
    }
};
